src/xrpl-api.js

```diff
@@ -116,29 +122,82 @@ class XrplApi {
+                for (let attempt = 0; attempt < attemptsPerServer;) {
+                    if (this.#isPrimaryServerConnected) {
+                        break serverIterator;
+                    }
+                    console.log(`Fallback server ${server} reconnection attempt ${++attempt}`);
+                    try {
+                        const client = new xrpl.Client(server, this.#xrplClientOptions);
+                        await client.connect();
+                        if (!this.#isPrimaryServerConnected) {
+                            this.#isFallbackServerConnected = true;server
+                            this.#client = client;
+                            console.log(`Successfully connected to the fallback server ${server}`)
+                        }
+                        break serverIterator;
+                    }
+                    catch (e) {
+                        console.log(`Error occurred while connecting to fallback server ${server}`, e)
+                        if (!this.#isPermanentlyDisconnected) {
+                            if(!maxRounds || round < maxRounds)
+                                console.log(`Fallback server ${server} connection attempt ${attempt} failed. Retrying in ${2 * round}s...`);
+                            else
+                                console.log(`Fallback server ${server} connection attempt failed.`);
+                            await new Promise(resolve => setTimeout(resolve, 2 * round * 1000));
+                        }
                     }
                 }
             }
+
+        }
+    }
+
+    async #attemptPrimaryServerReconnect() {
+        let attempt = 0;
+        while (!this.#isPermanentlyDisconnected && !this.#isPrimaryServerConnected) { // Keep attempting until consumer calls disconnect() manually.
+            console.log(`Primary server reconnection attempt ${++attempt}`);
+            try {
+                const client = new xrpl.Client(this.#primaryRippledServer, this.#xrplClientOptions);
+                await client.connect();
+                console.log("Successfully connected to the primary server");
+                this.#client = client;
+                this.#isPrimaryServerConnected = true;
+                break;
+            }
+            catch (e) {
+                console.log("Error occurred while re-connecting to the primary server", e)
+                if (!this.#isPermanentlyDisconnected) {
+                    const delaySec = 2 * attempt; // Retry with backoff delay.
+                    console.log(`Attempt ${attempt} failed. Retrying in ${delaySec}s...`);
+                    await new Promise(resolve => setTimeout(resolve, delaySec * 1000));
+                }
+            }
+        }
+    }
+
+    async #connectXrplClient(reconnect = false) {
+        if (reconnect) {
+            await Promise.all([this.#attemptFallbackServerReconnect(), this.#attemptPrimaryServerReconnect()]);
```

Comment:
As I understand from the code `this.#attemptPrimaryServerReconnect()` will keep running even if a fallback server is connected right? That's the requirement but I feel like there will be an issue in the following scenario.
1. Primary server disconnects
2. Then a fallback server is connected
3. Still the primary server connect promise is running
4. Then before the primary server gets up, the Previously connected fallback server disconnects
5. Now again this function is called from the disconnect event
6. Now another primary server connect promise is spawned while the previous promise is still running

Reply:
Another thing is when `this.#client` is replaced with all the `this.#client.on` events needed to be attached again. Otherwise, they won't be listened on replaced client.
I think it can be resolved if we change `async #initXrplClient()` -> `async #initXrplClient(client)` and call it with new client object. If so, you might need to change `this.#initXrplClient().then(() => this.#connectXrplClient(true));` line inside `this.#initXrplClient()` accordingly.
OR another solution is you might need to separate out initiating listeners to a separate function and call it accordingly.

Reply:
Another concern is let's say the client is connected to a fallback server. Then let's say there's a transaction happening. While the transaction is going on suddenly primary server comes up. Then the client will be replaced with the primary server. Then I feel that the ongoing transaction will fail due to this replacement.